content/software-testing-the-hardware-software-interface-part-2.md

```diff
@@ -0,0 +1,417 @@
+date: 2024-11-03
+author: Leonardo Held, Torizon Developer
+draft: false
+
+tags: ["Testing", "Embedded", "Linux", "Embedded"]
+
+abstract: A series about Embedded Software Testing.
+
+image: /soave-gate.jpg
+---
+
+This is the second installment in the embedded software testing series. This one will be about the architecture of a 
+system to test embedded software. We'll briefly cover the what is usually called a "dual-targetting" approach as well.
+
+### Theory
+
+In the previous installment we talked about the different levels of software testing and the last one was "end-to-end
+testing", which is treating the hardware or software Under Test (UDT) like a black box, poking it with a certain stimuli
+(inputs), seeing how it reacts (output) and comparing with a know expected value.
+
+So essentially we need something to actually do the poking, watch the outputs and compare with the known good values.
+Practically speaking though, that's hooking up UARTs, displays and whatever else to an embedded system and writing some
+software to do that automatically for us.
+
+```
++--------------------+             +-----------------------+
+|                    |             |                       |
+|  Testing Software  |             |   Device Under Test   |
+|   (Controller)     |             |       (DUT)           |
+|                    |             |     [Black Box]       |
+|                    |   poking    |                       |
+|  - Sends Inputs    | --------->  |  - Processes Inputs   |
+|    via UART,       |             |                       |
+|    Display, etc.   |             |                       |
+|                    |   reacting  |                       |
+|  - Receives        | <---------  |  - Generates Outputs  |
+|    Outputs         |             |    via UART, etc.     |
+|    via UART,       |             |                       |
+|    Other I/O       |             |                       |
++--------------------+             +-----------------------+
+```
+
+### Practice
+
+So, as an example let's assume you need to monitor whether your Linux box boots. The manual steps to achieve this are:
+
+1. Reset the board from whatever state it was
+2. Wait for the bootloader, kernel and userspace to boot
+3. See the login prompt
+
+Great, now we've reduced a requirement (the board must boot) to a set of manual execution steps. But you don't have just
+one board, you have around 100 different ones, and every single one of them must be tested every time your Continuous
+Integration builds a new OS for the board. Thinking about it in terms of automation, we can reduce this to:
+
+1. Connect the board's UART to a "Controller" - simply put, hooking up the board to a computer
+2. Write some software that parses the logs from the UART
+3. If it detects our login prompt, consider the boart fully booted
+
+As a practical example, I have a Verdin iMX8MP SoM connected to an Dahlia carrier board. Power is connected as is the
```

Comment:
"connected to an Dahlia" --> "connected to a Dahlia"

Reply:
Fixed, thanks.